base/compiler/ssair/inlining2.jl

```diff
@@ -26,271 +50,438 @@ function ssa_inlining_pass!(ir::IRCode, linetable::Vector{LineInfoNode}, sv::Opt
+        inline_compact = IncrementalCompact(compact, item.ir, compact.result_idx)
+        for (idx′, stmt′) in inline_compact
+            inline_compact[idx′] = nothing
+            stmt′ = ssa_substitute!(idx′, stmt′, argexprs, item.method.sig, item.sparams, linetable_offset, boundscheck_idx, compact)
+            if isa(stmt′, ReturnNode)
+                if isdefined(stmt′, :val)
+                    push!(pn.edges, inline_compact.active_result_bb-1)
+                    push!(pn.values, stmt′.val)
+                    stmt′ = GotoNode(post_bb_id)
+                end
+            elseif isa(stmt′, GotoNode)
+                stmt′ = GotoNode(stmt′.label + bb_offset)
+            elseif isa(stmt′, Expr) && stmt′.head == :enter
+                stmt′ = Expr(:enter, stmt′.args[1] + bb_offset)
+            elseif isa(stmt′, GotoIfNot)
+                stmt′ = GotoIfNot(stmt′.cond, stmt′.dest + bb_offset)
+            elseif isa(stmt′, PhiNode)
+                stmt′ = PhiNode(Any[edge+bb_offset for edge in stmt′.edges], stmt′.values)
+            end
+            inline_compact[idx′] = stmt′
+        end
+        just_fixup!(inline_compact)
+        compact.result_idx = inline_compact.result_idx
+        compact.active_result_bb = inline_compact.active_result_bb
+        for i = 1:length(pn.values)
+            isassigned(pn.values, i) || continue
+            if isa(pn.values[i], SSAValue)
+                compact.used_ssas[pn.values[i].id] += 1
+            end
+        end
+        if length(pn.edges) == 1
+            return_value = pn.values[1]
+        else
+            return_value = insert_node_here!(compact, pn, stmt.typ, compact.result_lines[idx])
+        end
+    end
+    return_value
+end
+
+function ir_inline_unionsplit!(compact, idx, argexprs, linetable, item, boundscheck, todo_bbs)
+    stmt, typ, line = compact.result[idx], compact.result_types[idx], compact.result_lines[idx]
+    atype = item.atype
+    generic_bb = item.bbs[end-1]
+    join_bb = item.bbs[end]
+    bb = compact.active_result_bb
+    pn = PhiNode()
+    has_generic = false
+    @assert length(item.bbs) > length(item.cases)
+    for ((metharg, case), next_cond_bb) in zip(item.cases, item.bbs)
+        case === nothing
+        @assert !isa(metharg, UnionAll)
+        cond = true
+        for (i, (a, m)) in Iterators.drop(enumerate(zip(atype.parameters, metharg.parameters)), 1)
+            # If this is always true, we don't need to check for it
+            a <: m && continue
+            # Generate isa check
+            isa_expr = Expr(:call, isa, argexprs[i], m)
+            isa_expr.typ = Bool
+            ssa = insert_node_here!(compact, isa_expr, Bool, line)
+            if cond === true
+                cond = ssa
+            else
+                and_expr = Expr(:call, and_int, cond, ssa)
+                cond = insert_node_here!(compact, and_expr, Bool, line)
+            end
+        end
+        atype = typesubtract(atype, metharg)
```

Comment:
attempting this operation is wrong, although since `atype` is a DataType, it's at least a no-op